File: ballast-core/dotnet/src/application/events/VesselMovedInDirection.cs
using System;
using Ballast.Core.Application.Models;
using Ballast.Core.Messaging;

namespace Ballast.Core.Application.Events
{
    public class VesselMovedInDirectionEvent : EventBase, IApplicationEvent  
    {

        public static string GetId() => nameof(VesselMovedInDirectionEvent);
        public override string Id => GetId();

        public Guid GameId { get; private set; }
        public VesselDto Vessel { get; private set; }
        public Direction Direction { get; private set; }

        private VesselMovedInDirectionEvent() { } // For model binding / deserialization
        private VesselMovedInDirectionEvent(string eventDateIsoString, Guid gameId, VesselDto vessel, Direction direction) : base(eventDateIsoString) 
        {
            GameId = gameId;
            Vessel = vessel;
            Direction = direction;
        }

        public static VesselMovedInDirectionEvent FromVesselDirectionInGame(Guid gameId, VesselDto vessel, Direction direction) =>
            new VesselMovedInDirectionEvent(
                EventBase.GetDateIsoString(),
                gameId, 
                vessel,
                direction
            );

    }
}

File: ballast-core/npm/src/application/events/vessel-moved-in-direction.ts
import { EventBase } from "../../messaging/event-base";
import { IApplicationEvent } from "../application-event";
import { IVesselDto } from "../models/vessel";
import { IDirection } from "../models/direction";

export interface IVesselMovedInDirectionEvent extends IApplicationEvent {
    readonly gameId: string;
    readonly vessel: IVesselDto;
    readonly direction: IDirection;
}

export class VesselMovedInDirectionEvent extends EventBase implements IVesselMovedInDirectionEvent {

    public static readonly id: string = "VesselMovedInDirectionEvent";
    public get id() {
        return VesselMovedInDirectionEvent.id;
    }

    public readonly gameId: string; 
    public readonly vessel: IVesselDto;
    public readonly direction: IDirection;

    private constructor(eventDateIsoString: string, gameId: string, vessel: IVesselDto, direction: IDirection) {
        super(eventDateIsoString);
        this.gameId = gameId;
        this.vessel = vessel;
        this.direction = direction;
    }

    public static fromJSON(json: IVesselMovedInDirectionEvent): VesselMovedInDirectionEvent {
        if (!json) {
            throw new Error("No json object was provided");
        }
        if (!json.id) {
            throw new Error("Missing id");
        }
        if (json.id != VesselMovedInDirectionEvent.id) {
            throw new Error("Id does not match event key");
        }
        if (!json.eventDateIsoString) {
            throw new Error("Missing eventDateIsoString");
        }
        if (!json.gameId) {
            throw new Error("Missing gameId");
        }
        if (!json.vessel) {
            throw new Error("Missing vessel");
        }
        if (!json.direction) {
            throw new Error("Missing vessel");
        }
        return new VesselMovedInDirectionEvent(json.eventDateIsoString, json.gameId, json.vessel, json.direction);
    }

    public static fromVesselDirectionInGame(gameId: string, vessel: IVesselDto, direction: IDirection) {
        return new VesselMovedInDirectionEvent(
            EventBase.getDateIsoString(),
            gameId,
            vessel,
            direction
        );
    }

}